gui/js/window_manager.js

```diff
@@ -77,11 +78,82 @@ module.exports = class WindowManager {
     return ''
   }
 
+  centerOnScreen (wantedWidth, wantedHeight) {
+    const bounds = this.win.getBounds()
+    // TODO : be smarter about which display to use ?
+    const display = electron.screen.getDisplayMatching(bounds)
+    const displaySize = display.workArea
+    const actualWidth = Math.min(wantedWidth, Math.floor(0.9 * displaySize.width))
+    const actualHeight = Math.min(wantedHeight, Math.floor(0.9 * displaySize.height))
+    this.win.setBounds({
+      x: Math.floor((displaySize.width - actualWidth) / 2),
+      y: Math.floor((displaySize.height - actualHeight) / 2),
+      width: actualWidth,
+      height: actualHeight
+    }, true /* animate on MacOS */)
+  }
+
+  placeWithTray (wantedWidth, wantedHeight, trayposition) {
+    const bounds = this.win.getBounds()
+    // TODO : be smarter about which display to use ?
+    const displayObject = electron.screen.getDisplayMatching(bounds)
+    const workArea = displayObject.workArea
+    const display = displayObject.bounds
+    const actualWidth = Math.min(wantedWidth, Math.floor(0.9 * workArea.width))
+    const actualHeight = Math.min(wantedHeight, Math.floor(0.9 * workArea.height))
+
+    const newBounds = {width: actualWidth, height: actualHeight}
+
+    if (process.platform === 'darwin') {
+      // on MacOS, try to center the popup below the tray icon
+      // later we might add a caret.
+
+      if (!trayposition || !trayposition.x) {
+        trayposition = {
+          x: workArea.width,
+          y: workArea.height
+        }
+      }
+
+      const centeredOnIcon = trayposition.x + Math.floor(trayposition.width / 2 - actualWidth / 2)
+      const fullRight = workArea.width - actualWidth
+
+      // in case where the icon is closer to the left border than half the windows width
+      // we put the window against the right border.
+      // later the caret will need to be moved from its center position in this case.
+      newBounds.x = Math.min(centeredOnIcon, fullRight)
+      newBounds.y = workArea.y // at the top
+
+    // all others OS let users define where to put the traybar
+    // icons are always on the right or bottom of the bar
+    // Let's try to guess where the bar is so we can place the window
+    // on window it is not platform-like to center above tray icon
```

Comment:
Did you write all this positioning logic? Or is it vendored from electron-positioner? With changes?

Reply:
Based on electron-positioner, but traybar detection is our own.

- [ ] `TODO`: contribute this to electron-positioner?